net: buf: Move net_buf_pool objects to dedicated linker area

jhedberg opened the pull request:
Moving the net_buf_pool objects to a dedicated area lets us access
them by array offset into this area instead of directly by pointer.
This helps reduce the size of net_buf objects by 4 bytes.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
net: buf: Move net_buf_pool objects to dedicated linker area

Moving the net_buf_pool objects to a dedicated area lets us access
them by array offset into this area instead of directly by pointer.
This helps reduce the size of net_buf objects by 4 bytes.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jukkar commented on include/net/buf.h:
```diff
@@ -590,7 +601,9 @@ struct net_buf *net_buf_get(struct k_fifo *fifo, s32_t timeout);
 static inline void net_buf_destroy(struct net_buf *buf)
 {
-	k_lifo_put(&buf->pool->free, buf);
+	struct net_buf_pool *pool = net_buf_pool_get(buf->pool_id);
```
If we are not checking the return value of net_buf_pool_get(), then why not write

k_lifo_put(&net_buf_pool_get(buf->pool_id)->free, buf)

instead and save extra variable here.

jhedberg replied:
No other reason except code clarity. Presumably the compiler will generate the same code either way with the default optimization level (I'd need to verify that however).

jukkar commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1055,9 +1055,12 @@ void bt_conn_recv(struct bt_conn *conn, struct net_buf *buf, u8_t flags)
 
-	if (buf->pool->user_data_size < BT_BUF_USER_DATA_MIN) {
+	pool = net_buf_pool_get(buf->pool_id);
+	if (pool->user_data_size < BT_BUF_USER_DATA_MIN) {
```
Same comment here, the pool variable looks useless.

jukkar reviewed (commented):
Minor comment, I am also fine with what you propose.

jukkar committed:
net: pkt: Fix net_pkt_split()

The net_pkt_split() was incorrectly checking fragA pointer
even before it was allocated.

The unit test is fixed and converted to ztest.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar reviewed (approved):
LGTM
